File: TypeScript/279. Perfect Squares.ts
const numSquares = (n: number): number => {
    const sq: number[] = [...Array(n).keys()].map(num => num * num).filter(num => 0 < num && num <= n)
    const queue: { node: number, d: number }[] = [{node: n, d: 1}]
    const seen = new Set()
    seen.add(n)
    while (queue) {
        const curr = queue.shift()

        if (!curr) {
            break
        }

        const {node, d}: any = curr

        if (sq.indexOf(node) !== -1) {
            return d
        }

        for (let i = 0; i < sq.length; i++) {
            if (node <= sq[i]) {
                break
            }

            const left: number = node - sq[i]
            if (!seen.has(left)) {
                seen.add(left)
                queue.push({node: left, d: d + 1})
            }
        }
    }
    return n
}

File: Python/279. Perfect Squares.py
from collections import deque


class Solution:
    def numSquares(self, n: int) -> int:
        sq = [x * x for x in range(1, n) if x * x <= n]
        queue = deque([(n, 1)])
        seen = set()
        seen.add(n)
        while queue:
            node, d = queue.popleft()
            if node in sq:
                return d
            for s in sq:
                if node <= s:
                    break

                left = node - s
                if left not in seen:
                    seen.add(left)
                    queue.append((left, d + 1))
        return n
